### PR TITLE
Fix the cuda left-hand swizzle issue (#3538)

### PR DESCRIPTION
**Pull request by kaizhangNV**

### CODE REVIEW

**Comment by kaizhangNV** [comment]
The change basically only generate
`foo.xy = vec2 `
to

`foo.x = vec2.x`
`foo.y = vec2.y`

for C++ and CUDA target.

File: source/slang/slang-emit-c-like.cpp [hunk 1]
```diff
@@ -352,6 +352,50 @@ void CLikeSourceEmitter::_emitType(IRType* type, DeclaratorInfo* declarator)
...
 
+void CLikeSourceEmitter::_emitSwizzleStorePerElement(IRInst* inst)
+{
+    auto subscriptOuter = getInfo(EmitOp::General);
+    auto subscriptPrec = getInfo(EmitOp::Postfix);
+
+    // store the right-handside expression into a temporary variable first, if
+    // the right-handside expression is already a variable, we can skip this.
+    auto ii = cast<IRSwizzledStore>(inst);
+    String tempIdentifierName = "tmp";
```
Comment (csyonghe) [inline source/slang/slang-emit-c-like.cpp:363]: Don't do this yourself, instead, just make sure we don't fold `ii->getSource()` by handling a case for IRSwizzledStore use after line 1496 in slang-emit-c-like.cpp (inside CLikeSourceEmitter::shouldFoldInstIntoUseSites):

```
if (auto swizzledStore = as<IRSwizzledStore>(inst))
{
      if (isCPUTarget(getTargetReq()) || isCUDATarget(getTargetReq()))
           return false;
}
```

Comment (csyonghe) [reply source/slang/slang-emit-c-like.cpp:363]: This should make sure that we will always emit the source operand as a standalone assignment statement before we reach the swizzled store inst. So we can simply do a loop and call emitOperand(ii->getSource()) in the loop without duplicating the logic that computes `source`.

File: source/slang/slang-emit-c-like.cpp [hunk 2]
```diff
@@ -2760,32 +2804,44 @@ void CLikeSourceEmitter::_emitInst(IRInst* inst)
...
+            switch (this->getTarget())
+            {
+                case CodeGenTarget::CPPSource:
+                case CodeGenTarget::HostCPPSource:
+                case CodeGenTarget::CUDASource:
+                    // cpp and cuda target don't support swizzle on the left handside, so we
+                    // have to assign the element one by one.
+                    _emitSwizzleStorePerElement(inst);
+                    break;
+                default:
```
Comment (csyonghe) [inline source/slang/slang-emit-c-like.cpp:2816]: You can `return` for the three cases and keep the remaining logic outside of the `switch` to reduce nesting levels.
You can also just call `isCPUTarget(getTargetReq()) || isCUDATarget(getTargetReq())`.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**